pkg/services/sqlstore/datasource.go

```diff
@@ -120,39 +120,30 @@ func (ss *SQLStore) DeleteDataSource(uid string, id int64, name string, orgID in
 	return cmd.DeletedDatasourcesCount, nil
 }
 
-// DeleteDataSource removes a datasource by org_id as well as either uid (preferred), id, or name
-// and is added to the bus.
+// DeleteDataSource removes a datasource by org_id as well as uid
 func DeleteDataSource(cmd *models.DeleteDataSourceCommand) error {
-	params := make([]interface{}, 0)
-
-	makeQuery := func(sql string, p ...interface{}) {
-		params = append(params, sql)
-		params = append(params, p...)
-	}
+	return inTransaction(func(sess *DBSession) error {
+		query := &models.GetDataSourceQuery{
+			Id:    cmd.ID,
+			Uid:   cmd.UID,
+			Name:  cmd.Name,
+			OrgId: cmd.OrgID,
+		}
 
-	switch {
-	case cmd.OrgID == 0:
-		return models.ErrDataSourceIdentifierNotSet
-	case cmd.UID != "":
-		makeQuery("DELETE FROM data_source WHERE uid=? and org_id=?", cmd.UID, cmd.OrgID)
-	case cmd.ID != 0:
-		makeQuery("DELETE FROM data_source WHERE id=? and org_id=?", cmd.ID, cmd.OrgID)
-	case cmd.Name != "":
-		makeQuery("DELETE FROM data_source WHERE name=? and org_id=?", cmd.Name, cmd.OrgID)
-	default:
-		return models.ErrDataSourceIdentifierNotSet
-	}
+		if err := GetDataSource(query); err != nil {
+			cmd.DeletedDatasourcesCount = 0
+			return nil
+		}
 
-	return inTransaction(func(sess *DBSession) error {
-		result, err := sess.Exec(params...)
+		result, err := sess.Exec("DELETE FROM data_source WHERE uid=? and org_id=?", query.Result.Uid, query.Result.OrgId)
 		cmd.DeletedDatasourcesCount, _ = result.RowsAffected()
 
 		sess.publishAfterCommit(&events.DataSourceDeleted{
```

Comment:
We probably only want to publish the event if `err != nil`

Reply:
updated to handle the errors from `sess.Exec` and `result.RowsAffected()`